public/content/en/gems/traits.md

```diff
@@ -0,0 +1,169 @@
+`commonPrefix` differentiates between `RandomAccessRange`s
+and linear iterable ranges, because in `RandomAccessRange` it's possible to jump
+between positions and thus speed-up the algorithm.
+
+#### More CTFE magic
+
+[std.traits](https://dlang.org/phobos/std_traits.html) wraps most of
+D's [traits](https://dlang.org/spec/traits.html) except for some like
+`compiles` that can't be wrapped as it would lead to an immediate compile error:
+
+```
+__traits(compiles, obvious error - $%42); // false
+```
+
+#### Special keywords
+
+Additionally for debugging purposes D provides a couple of special keywords:
+
+```
+void test(string file = __FILE__, size_t line = __LINE__, string mod = __MODULE__,
+          string func = __FUNCTION__, string pretty = __PRETTY_FUNCTION__)
+{
+    writefln("file: '%s', line: '%s', module: '%s',\nfunction: '%s', pretty function: '%s'",
+             file, line, mod, func, pretty);
+}
+```
+
+With D's CLI evaluation one doesn't even need `time` - CTFE can be used!
+
+```
+rdmd --force --eval='pragma(msg, __TIMESTAMP__);'
+```
+
+## In-depth
+
+- [std.range.primitives](https://dlang.org/phobos/std_range_primitives.html)
+- [std.traits](https://dlang.org/phobos/std_traits.html)
+- [std.meta](https://dlang.org/phobos/std_meta.html)
+- [Specification on Traits in D](https://dlang.org/spec/traits.html)
+
+## {SourceCode}
+
+import std.functional: binaryFun;
+import std.range;
+import std.stdio;
+import std.traits;
+
+/**
+Returns the common prefix of two ranges
+without the auto-decoding special case.
+
+Params:
+    pred = Predicate for commonality comparison
+    r1 = A forward range of elements.
+    r2 = An input range of elements.
+
+Returns:
+A slice of r1 which contains the characters
+that both ranges start with.
+ */
+auto commonPrefix(alias pred = "a == b", R1, R2)
+                 (R1 r1, R2 r2)
+if (isForwardRange!R1 && isInputRange!R2 &&
+    !isNarrowString!R1 &&
+    is(typeof(binaryFun!pred(r1.front,
+                             r2.front))))
+{
+    import std.algorithm.comparison : min;
+    static if (isRandomAccessRange!R1 &&
+               isRandomAccessRange!R2 &&
+               hasLength!R1 && hasLength!R2 &&
+               hasSlicing!R1)
+    {
+        immutable limit = min(r1.length,
+                              r2.length);
+        foreach (i; 0 .. limit)
+        {
+            if (!binaryFun!pred(r1[i], r2[i]))
+            {
+                return r1[0 .. i];
+            }
+        }
+        return r1[0 .. limit];
+    }
+    else
+    {
+        import std.range : takeExactly;
+        auto result = r1.save;
+        size_t i = 0;
+        for (;
+             !r1.empty && !r2.empty &&
+             binaryFun!pred(r1.front, r2.front);
+             ++i, r1.popFront(), r2.popFront())
+        {}
+        return takeExactly(result, i);
+    }
+}
+
+void main()
+{
+    // prints: "hello, "
+    writeln(commonPrefix("hello, world"d,
```

Comment:
Well the problem here is that narrow strings require autodecoding in D (if used with ranges and without workarounds) and this example here is explicitly about random access.
I guess I could 
a) search for a better example
b) change from string to number (probably a bit boring)